table/tables/tables.go

```diff
@@ -187,114 +213,102 @@ func (t *Table) FirstKey() kv.Key {
 }
 
 // UpdateRecord implements table.Table UpdateRecord interface.
-func (t *Table) UpdateRecord(ctx context.Context, h int64, oldData []types.Datum, newData []types.Datum, touched map[int]bool) error {
-	// We should check whether this table has on update column which state is write only.
-	currentData := make([]types.Datum, len(t.WritableCols()))
-	copy(currentData, newData)
+func (t *Table) UpdateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, touched []bool) error {
+	txn := ctx.Txn()
+	bs := kv.NewBufferStore(txn)
 
-	// If they are not set, and other data are changed, they will be updated by current timestamp too.
-	err := t.setOnUpdateData(ctx, touched, currentData)
+	// rebuild index
+	err := t.rebuildIndices(bs, h, touched, oldData, newData)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	txn := ctx.Txn()
-	bs := kv.NewBufferStore(ctx.Txn())
-
-	// Compose new row
-	t.composeNewData(touched, currentData, oldData)
-	colIDs := make([]int64, 0, len(t.WritableCols()))
-	for i, col := range t.WritableCols() {
-		if col.State != model.StatePublic && currentData[i].IsNull() {
-			defaultVal, err1 := table.GetColDefaultValue(ctx, col.ToInfo())
-			if err1 != nil {
-				return errors.Trace(err1)
+	var colIDs, binlogColIDs []int64
+	var row, binlogOldRow, binlogNewRow []types.Datum
+	colIDs = make([]int64, 0, len(newData))
+	row = make([]types.Datum, 0, len(newData))
+	if shouldWriteBinlog(ctx) {
+		binlogColIDs = make([]int64, 0, len(newData))
+		binlogOldRow = make([]types.Datum, 0, len(newData))
+		binlogNewRow = make([]types.Datum, 0, len(newData))
+	}
+
+	for _, col := range t.WritableCols() {
+		var value types.Datum
+		if col.State != model.StatePublic {
+			// If col is in write only or write reorganization state
+			// and the value is not default, keep the original value.
+			value, err = table.GetColOriginDefaultValue(ctx, col.ToInfo())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			cmp, err := oldData[col.Offset].CompareDatum(ctx.GetSessionVars().StmtCtx, value)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if cmp != 0 {
+				value = oldData[col.Offset]
 			}
-			currentData[i] = defaultVal
+		} else {
+			value = newData[col.Offset]
+		}
+		if !t.canSkip(col, value) {
+			colIDs = append(colIDs, col.ID)
+			row = append(row, value)
+		}
+		if shouldWriteBinlog(ctx) && !t.canSkipUpdateBinlog(col, value) {
+			binlogColIDs = append(binlogColIDs, col.ID)
```

Comment:
Because `colIDs` is for encoding values into row **stored in Tikv**, but `binlogColIDs` is used for **binlog**. They are differant. For example, if a column is PK handle, it won't be encoded into row for saving space, but it should appear in binlog.